flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTaskTest.java

```diff
@@ -479,5 +481,40 @@ public void cancel() {
 			running = false;
 		}
 	}
+
+	private static final class OutputRecordInCloseTestSource extends RichSourceFunction<String> {
+
+		private final String name;
+		private final String[] elements;
+
+		private transient volatile boolean canceled = false;
+
+		private transient SourceContext<String> context;
+
+		public OutputRecordInCloseTestSource(String name, String ... elements) {
+			this.name = name;
+			this.elements = elements.clone();
+		}
+
+		public void run(SourceContext<String> ctx) {
+			context = ctx;
+			int i = 0;
+			while (!canceled && i < elements.length) {
+				context.collect(elements[i++]);
+			}
+		}
+
+		public void close() {
+			if (!canceled) {
+				checkNotNull(context);
+				context.collect("[" + name + "]: Bye");
+			}
+		}
+
+		@Override
+		public void cancel() {
+			canceled = true;
```

Comment:
I guess we could make this method empty and remove `canceled` variable for this concrete unit test.